source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+
+      // For the next part of the trie, the next available address is at address + 2^branching
+      // factor.
+      *nextfree += 1 << output.branch_;
+      p = first;
+
+      // Build the subtrees.
+      for (bit_pattern = 0; bit_pattern < static_cast<uint32_t>(1 << output.branch_);
+           ++bit_pattern) {
+
+        k = 0;
+        // k is finding the number of entries in the base vector that have the same bit pattern.
+        while (p + k < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[p + k].ip_)) == bit_pattern) {
+          ++k;
+        }
+
+        if (k == 0) {
+          if (p == first + n) {
+            buildRecursive(output.prefix_ + output.branch_, p - 1, 1, address + bit_pattern,
+                           nextfree);
+          } else {
+            buildRecursive(output.prefix_ + output.branch_, p, 1, address + bit_pattern, nextfree);
+          }
+        } else if (k == 1 && ip_prefixes_[p].length_ - output.prefix_ < output.branch_) {
+          uint32_t i;
+          bits = output.branch_ + output.prefix_ - ip_prefixes_[p].length_;
+          for (i = bit_pattern; i < bit_pattern + (1 << bits); ++i) {
+            buildRecursive(output.prefix_ + output.branch_, p, 1, address + i, nextfree);
+          }
+        } else {
+          buildRecursive(output.prefix_ + output.branch_, p, k, address + bit_pattern, nextfree);
+        }
+        p += k;
+      }
+    }
+
+    /**
+     * LcNode is a uint32_t. A wrapper is provided to simplify getting the branch, skip and
+     * address values held within the data structure.
+     *
+     * The LcNode has three parts to it
+     * - Branch: the first 5 bits represent the branching factor. This supports a branching factor
+     * of 2^31.
+     * - Skip: the next 7 bits represent the number of bits to skip when looking at an IP address.
+     * This value can be between 0 and 127, so IPv6 is supported as well.
+     * - Address: the remaining 30 bits represent the pointer. If branch value is not zero, this
+     * address points to another entry in the trie. If branch is zero, the address points to an
+     * entry in the base vector.
+     */
+    typedef uint32_t LcNode;
+    /**
+     * Takes a branch value and shifts it into the first 5 bits of uint32_t.
+     */
+    static LcNode setBranch(LcNode node) { return node << 27; }
+
+    /**
+     * @return Returns the first 5 bits of LcNode.
+     */
+    static LcNode getBranch(LcNode node) { return node >> 27; }
+
+    /**
+     * Takes a skip value and shifts it itno the 6-12(inclusively) bits of uint32_t.
+     * @param node
+     * @return
+     */
+    static LcNode setSkip(LcNode node) { return node << 20; }
+
+    /**
+     * @return Returns the value at bits 6-12.
+     */
+    static LcNode getSkip(LcNode node) { return node >> 20 & 127; }
```

Comment:
In order to make code like this error-prone, the mechanism I usually use is:
```
struct LcNode {
  uint32_t branch:5;
  uint32_t skip:7;
  uint32_t address:20;
};
```
Then you can just refer to fields in the struct by name, and let the compiler deal with all the bit shifting.

I don't know if bitfield structs like that are encouraged or discouraged in Envoy, though.

Reply:
this is fine, SGTM.

Reply:
Thank you for the suggestion. This is cleaner. I've updated the code to use your proposed structure.